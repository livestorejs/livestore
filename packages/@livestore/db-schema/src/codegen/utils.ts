import os from 'node:os'
import path from 'node:path'

import dprint from 'dprint-node'

export const pretty = (code: string): string => {
  const tmpPath = os.tmpdir()
  const filePath = path.join(tmpPath, 'tmp.ts')

  return dprint.format(filePath, code, {
    semiColons: 'asi',
    quoteStyle: 'preferSingle',
  })
}
