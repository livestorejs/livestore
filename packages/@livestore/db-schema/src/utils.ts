export type Prettify<T> = T extends infer U ? { [K in keyof U]: Prettify<U[K]> } : never
export type PrettifyFlat<T> = T extends infer U ? { [K in keyof U]: U[K] } : never

export type Nullable<T> = { [K in keyof T]: T[K] | null }
