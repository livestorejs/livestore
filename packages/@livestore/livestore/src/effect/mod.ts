export {
  DeferredStoreContext,
  type LiveStoreContextProps,
  LiveStoreContextRunning as LiveStoreContext,
  LiveStoreContextRunning,
} from '../store/create-store.ts'
export { LiveStoreContextDeferred, LiveStoreContextLayer } from './LiveStore.ts'
