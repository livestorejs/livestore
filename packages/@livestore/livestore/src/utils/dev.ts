/* eslint-disable unicorn/prefer-global-this */
export const downloadBlob = (
  data: Uint8Array | Blob | string,
  fileName: string,
  mimeType = 'application/octet-stream',
) => {
  const blob = data instanceof Blob ? data : new Blob([data], { type: mimeType })

  const url = window.URL.createObjectURL(blob)

  downloadURL(url, fileName)

  setTimeout(() => window.URL.revokeObjectURL(url), 1000)
}

export const downloadURL = (data: string, fileName: string) => {
  const a = document.createElement('a')
  a.href = data
  a.download = fileName
  document.body.append(a)
  a.style.display = 'none'
  a.click()
  a.remove()
}

export const exposeDebugUtils = () => {
  if (import.meta.env.DEV) {
    globalThis.__debugDownloadBlob = downloadBlob
  }
}
