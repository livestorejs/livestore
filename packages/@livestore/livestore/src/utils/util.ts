/// <reference lib="es2022" />

import type { Brand } from '@livestore/utils/effect'

export type ParamsObject = Record<string, SqlValue>
export type SqlValue = string | number | Uint8Array | null

export type Bindable = SqlValue[] | ParamsObject

export type PreparedBindValues = Brand.Branded<Bindable, 'PreparedBindValues'>

/**
 * This is a tag function for tagged literals.
 * it lets us get syntax highlighting on SQL queries in VSCode, but
 * doesn't do anything at runtime.
 * Code copied from: https://esdiscuss.org/topic/string-identity-template-tag
 */
export const sql = (template: TemplateStringsArray, ...args: unknown[]): string => {
  let str = ''

  for (const [i, arg] of args.entries()) {
    str += template[i] + String(arg)
  }

  // eslint-disable-next-line unicorn/prefer-at
  return str + template[template.length - 1]
}

/** Prepare bind values to send to SQLite
/*  Add $ to the beginning of keys; which we use as our interpolation syntax
/*  We also strip out any params that aren't used in the statement,
/*  because rusqlite doesn't allow unused named params
/*  TODO: Search for unused params via proper parsing, not string search
**/
export const prepareBindValues = (values: Bindable, statement: string): PreparedBindValues => {
  if (Array.isArray(values)) return values as PreparedBindValues

  const result: ParamsObject = {}
  for (const [key, value] of Object.entries(values)) {
    if (statement.includes(key)) {
      result[`$${key}`] = value
    }
  }

  return result as PreparedBindValues
}

/**
 * Use this to make assertion at end of if-else chain that all members of a
 * union have been accounted for.
 */
/* eslint-disable-next-line prefer-arrow/prefer-arrow-functions */
export function casesHandled(x: never): never {
  throw new Error(`A case was not handled for value: ${objectToString(x)}`)
}

export const objectToString = (error: any): string => {
  const stack = typeof process !== 'undefined' && process.env.CL_DEBUG ? error.stack : undefined
  const str = error.toString()
  const stackStr = stack ? `\n${stack}` : ''
  if (str !== '[object Object]') return str + stackStr

  try {
    return JSON.stringify({ ...error, stack }, null, 2)
  } catch (e: any) {
    console.log(error)

    return 'Error while printing error: ' + e
  }
}

export const isPromise = (value: any): value is Promise<unknown> => typeof value?.then === 'function'
