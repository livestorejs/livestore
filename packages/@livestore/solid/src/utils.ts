import type { Accessor } from 'solid-js'

export type MakeOptional<T, TKeys extends keyof T> = Omit<T, TKeys> & { [TKey in TKeys]: T[TKey] | undefined }

export type AccessorMaybe<T> = Accessor<T> | T

export function resolve<T>(value: AccessorMaybe<T>): T {
  if (typeof value === 'function') {
    return (value as Accessor<T>)()
  }
  return value
}
