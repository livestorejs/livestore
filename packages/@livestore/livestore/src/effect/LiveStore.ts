import type { Scope } from '@livestore/utils/effect'
import { Context, Deferred, Duration, Effect, Layer, OtelTracer, pipe, Runtime } from '@livestore/utils/effect'
import * as otel from '@opentelemetry/api'
import type { GraphQLSchema } from 'graphql'

// import initSqlite3Wasm from 'sqlite-esm'
import type { InMemoryDatabase } from '../inMemoryDatabase.js'
import type { LiveStoreSchema } from '../schema/index.js'
import type { StorageInit } from '../storage/index.js'
import type { BaseGraphQLContext, GraphQLOptions, LiveStoreQuery, Store } from '../store.js'
import { createStore } from '../store.js'

// NOTE we're starting to initialize the sqlite wasm binary here (already before calling `createStore`),
// so that it's ready when we need it
// const sqlite3Promise = initSqlite3Wasm({
//   print: (message) => console.log(`[livestore sqlite] ${message}`),
//   printErr: (message) => console.error(`[livestore sqlite] ${message}`),
// })

// TODO get rid of `LiveStoreContext` wrapper and only expose the `Store` directly
export type LiveStoreContext = {
  store: Store
}

export type QueryDefinition = (store: Store) => LiveStoreQuery

type ExecThisOptions = any
type ExecResultRowsOptions = any

// type PreparedStatement = {
//   // db: DatabaseApi
//   // bind(value: any): PreparedStatement
//   // bind(idx: number, value: any): PreparedStatement
//   // bindAsBlob(value: any): any
//   // bindAsBlob(idx: number, value: any): any
//   // get(ndx: number, asType?: any): any

//   // finalize(): any
//   // stepFinalize(): boolean

//   // // TODO
//   // columnCount(): number
//   // parameterCount(): number
//   // clearBindings(): any
//   reset(): any
//   step(): any
//   // stepReset(): any
//   // getInt(c: number): number
//   // getFloat(c: number): number
//   // getString(c: number): string
//   // getBlob(c: number): Uint8Array
//   // getJSON: any
//   // getColumnName(c: number): string
//   getColumnNames(): string[]
//   // getParamIndex: any
//   // pointer: number
// }

export type SQLiteBindValue = string | number | null | boolean | Uint8Array
export type SQLiteBindParams = Record<string, SQLiteBindValue> | SQLiteBindValue[]
export type SQLiteVariadicBindParams = SQLiteBindValue[]

export type SQLiteBindPrimitiveParams = Record<string, Exclude<SQLiteBindValue, Uint8Array>>
export type SQLiteBindBlobParams = Record<string, Uint8Array>
export type SQLiteColumnNames = string[]
export type SQLiteColumnValues = any[]
export type SQLiteAnyDatabase = any

export interface SQLiteExecuteSyncResult<T> extends IterableIterator<T> {
  /**
   * The last inserted row ID. Returned from the [`sqlite3_last_insert_rowid()`](https://www.sqlite.org/c3ref/last_insert_rowid.html) function.
   */
  readonly lastInsertRowId: number

  /**
   * The number of rows affected. Returned from the [`sqlite3_changes()`](https://www.sqlite.org/c3ref/changes.html) function.
   */
  readonly changes: number

  /**
   * Get the first row of the result set. This requires the SQLite cursor to be in its initial state. If you have already retrieved rows from the result set, you need to reset the cursor first by calling [`resetSync()`](#resetsync). Otherwise, an error will be thrown.
   */
  getFirstSync(): T | null

  /**
   * Get all rows of the result set. This requires the SQLite cursor to be in its initial state. If you have already retrieved rows from the result set, you need to reset the cursor first by calling [`resetSync()`](#resetsync). Otherwise, an error will be thrown.
   */
  getAllSync(): T[]

  /**
   * Reset the prepared statement cursor. This will call the [`sqlite3_reset()`](https://www.sqlite.org/c3ref/reset.html) C function under the hood.
   */
  resetSync(): void
}

type ValuesOf<T extends object> = T[keyof T][]

export interface PreparedStatement {
  /**
   * Run the prepared statement and return the [`SQLiteExecuteSyncResult`](#sqliteexecutesyncresult) instance.
   * > **Note:** Running heavy tasks with this function can block the JavaScript thread and affect performance.
   * @param params The parameters to bind to the prepared statement. You can pass values in array, object, or variadic arguments. See [`SQLiteBindValue`](#sqlitebindvalue) for more information about binding values.
   */
  executeSync<T>(params: SQLiteBindParams): SQLiteExecuteSyncResult<T>
  /**
   * @hidden
   */
  executeSync<T>(...params: SQLiteVariadicBindParams): SQLiteExecuteSyncResult<T>
  /**
   * Similar to [`executeSync()`](#executesyncparams) but returns the raw value array result instead of the row objects.
   * @hidden Advanced use only.
   */
  executeForRawResultSync<T extends object>(params: SQLiteBindParams): SQLiteExecuteSyncResult<ValuesOf<T>>
  /**
   * @hidden
   */
  executeForRawResultSync<T extends object>(...params: SQLiteVariadicBindParams): SQLiteExecuteSyncResult<ValuesOf<T>>
  /**
   * Get the column names of the prepared statement.
   */
  getColumnNamesSync(): string[]
  /**
   * Finalize the prepared statement. This will call the [`sqlite3_finalize()`](https://www.sqlite.org/c3ref/finalize.html) C function under the hood.
   *
   * Attempting to access a finalized statement will result in an error.
   * > **Note:** While expo-sqlite will automatically finalize any orphaned prepared statements upon closing the database, it is considered best practice to manually finalize prepared statements as soon as they are no longer needed. This helps to prevent resource leaks. You can use the `try...finally` statement to ensure that prepared statements are finalized even if an error occurs.
   */
  finalizeSync(): void
}

export type DatabaseApi = {
  filename: string
  pointer: number
  exec(input: string, opts?: ExecThisOptions): DatabaseApi
  exec(input: string, opts?: ExecResultRowsOptions): any

  exec(opts: ExecThisOptions): DatabaseApi
  exec(opts: ExecResultRowsOptions): any

  // exec(opts: ExecOptions & {returnValue: "resultRows"}): any;
  prepare(sql: string): PreparedStatement

  isOpen: () => boolean
  affirmOpen: () => DatabaseApi
  close: () => void
  changes: (total?: boolean, sixtyFour?: boolean) => number
  dbFilename: () => string
  dbName: () => string
  dbVfsName: (dbName: any) => string
  createFunction: Function

  selectValue: Function
  selectValues: Function
  selectArray: Function
  selectObject: Function
  selectArrays: Function
  selectObjects: Function

  openStatementCount: Function
  transaction: Function
  savepoint: Function
  checkRc: Function
}

export type LiveStoreCreateStoreOptions<GraphQLContext extends BaseGraphQLContext> = {
  schema: LiveStoreSchema
  loadStorage: () => StorageInit | Promise<StorageInit>
  graphQLOptions?: GraphQLOptions<GraphQLContext>
  otelTracer?: otel.Tracer
  otelRootSpanContext?: otel.Context
  boot?: (db: InMemoryDatabase, parentSpan: otel.Span) => unknown | Promise<unknown>
  sqlite3: DatabaseApi
}

export const LiveStoreContext = Context.Tag<LiveStoreContext>('@livestore/livestore/LiveStoreContext')

export type DeferredStoreContext = Deferred.Deferred<never, LiveStoreContext>
export const DeferredStoreContext = Context.Tag<DeferredStoreContext>(
  Symbol.for('@livestore/livestore/DeferredStoreContext'),
)

// export const DeferredStoreContext = Effect.cached(Effect.flatMap(StoreContext, (_) => Effect.succeed(_)))

export type LiveStoreContextProps<GraphQLContext extends BaseGraphQLContext> = {
  schema: LiveStoreSchema
  loadStorage: () => StorageInit | Promise<StorageInit>
  graphQLOptions?: {
    schema: Effect.Effect<otel.Tracer, never, GraphQLSchema>
    makeContext: (db: InMemoryDatabase) => GraphQLContext
  }
  boot?: (db: InMemoryDatabase) => Effect.Effect<never, never, void>
}

export const LiveStoreContextLayer = <GraphQLContext extends BaseGraphQLContext>(
  props: LiveStoreContextProps<GraphQLContext>,
): Layer.Layer<otel.Tracer, never, LiveStoreContext> =>
  Layer.scoped(LiveStoreContext, makeLiveStoreContext(props)).pipe(
    Layer.withSpan('LiveStore'),
    Layer.provide(LiveStoreContextDeferred),
  )

export const LiveStoreContextDeferred = Layer.effect(DeferredStoreContext, Deferred.make<never, LiveStoreContext>())

export const makeLiveStoreContext = <GraphQLContext extends BaseGraphQLContext>({
  schema,
  loadStorage,
  graphQLOptions: graphQLOptions_,
  boot: boot_,
}: LiveStoreContextProps<GraphQLContext>): Effect.Effect<
  DeferredStoreContext | Scope.Scope | otel.Tracer,
  never,
  LiveStoreContext
> =>
  pipe(
    Effect.gen(function* ($) {
      const runtime = yield* $(Effect.runtime<never>())

      const otelRootSpanContext = otel.context.active()

      const otelTracer = yield* $(OtelTracer.Tracer)

      const graphQLOptions = yield* $(
        graphQLOptions_
          ? Effect.all({ schema: graphQLOptions_.schema, makeContext: Effect.succeed(graphQLOptions_.makeContext) })
          : Effect.succeed(undefined),
      )

      const boot = boot_
        ? (db: InMemoryDatabase) =>
            boot_(db).pipe(Effect.withSpan('boot'), Effect.tapCauseLogPretty, Runtime.runPromise(runtime))
        : undefined

      // @ts-expect-error
      const sqlite3 = yield* $(Effect.promise(() => sqlite3Promise))

      const store = yield* $(
        Effect.tryPromise(() =>
          createStore({
            schema,
            loadStorage,
            graphQLOptions,
            otelTracer,
            otelRootSpanContext,
            boot,
            // @ts-expect-error
            sqlite3,
          }),
        ),
        Effect.acquireRelease((store) => Effect.sync(() => store.destroy())),
      )

      window.__debugLiveStore = store

      return { store }
    }),
    Effect.tap((storeCtx) => Effect.flatMap(DeferredStoreContext, (def) => Deferred.succeed(def, storeCtx))),
    Effect.timeoutFail({
      onTimeout: () => new Error('Timed out while creating LiveStore store after 10sec'),
      duration: Duration.seconds(10),
    }),
    Effect.orDie,
  )
