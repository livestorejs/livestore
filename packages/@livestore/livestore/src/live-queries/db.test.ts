import { Effect, Schema } from '@livestore/utils/effect'
import * as otel from '@opentelemetry/api'
import { BasicTracerProvider, InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import { describe, expect, it } from 'vitest'

import { computed, queryDb, rawSqlMutation, sql } from '../index.js'
import { makeTodoMvc, tables } from '../utils/tests/fixture.js'
import { getSimplifiedRootSpan } from '../utils/tests/otel.js'

/*
TODO write tests for:

- sql queries without and with `map` (incl. callback and schemas)
- optional and explicit `queriedTables` argument
*/

describe('otel', () => {
  let cachedProvider: BasicTracerProvider | undefined

  const makeQuery = Effect.gen(function* () {
    const exporter = new InMemorySpanExporter()

    const provider = cachedProvider ?? new BasicTracerProvider()
    cachedProvider = provider
    provider.addSpanProcessor(new SimpleSpanProcessor(exporter))
    provider.register()

    const otelTracer = otel.trace.getTracer('test')

    const span = otelTracer.startSpan('test')
    const otelContext = otel.trace.setSpan(otel.context.active(), span)

    const { store } = yield* makeTodoMvc({ otelTracer, otelContext })

    return {
      store,
      otelTracer,
      exporter,
      span,
      provider,
    }
  })

  it('otel', async () => {
    const { exporter } = await Effect.gen(function* () {
      const { store, exporter, span } = yield* makeQuery

      const query$ = queryDb({
        query: `select * from todos`,
        schema: Schema.Array(tables.todos.schema),
        queriedTables: new Set(['todos']),
      })
      expect(query$.run()).toMatchInlineSnapshot('[]')

      store.mutate(rawSqlMutation({ sql: sql`INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)` }))

      expect(query$.run()).toMatchInlineSnapshot(`
      [
        {
          "completed": false,
          "id": "t1",
          "text": "buy milk",
        },
      ]
    `)

      query$.destroy()
      span.end()

      return { exporter }
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise)

    expect(getSimplifiedRootSpan(exporter)).toMatchSnapshot()
  })

  it('with thunks', async () => {
    const { exporter } = await Effect.gen(function* () {
      const { store, exporter, span } = yield* makeQuery

      const defaultTodo = { id: '', text: '', completed: false }

      const filter = computed(() => `where completed = 0`, { label: 'where-filter' })
      const query$ = queryDb(
        (get) => ({
          query: `select * from todos ${get(filter)}`,
          schema: Schema.Array(tables.todos.schema).pipe(Schema.headOrElse(() => defaultTodo)),
        }),
        { label: 'all todos' },
      )

      expect(query$.run()).toMatchInlineSnapshot(`
      {
        "completed": false,
        "id": "",
        "text": "",
      }
    `)

      store.mutate(rawSqlMutation({ sql: sql`INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)` }))

      expect(query$.run()).toMatchInlineSnapshot(`
      {
        "completed": false,
        "id": "t1",
        "text": "buy milk",
      }
    `)

      query$.destroy()
      span.end()

      return { exporter }
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise)

    expect(getSimplifiedRootSpan(exporter)).toMatchSnapshot()
  })

  it('with thunks with query builder and without labels', async () => {
    const { exporter } = await Effect.gen(function* () {
      const { store, exporter, span } = yield* makeQuery

      const defaultTodo = { id: '', text: '', completed: false }

      const filter = computed(() => ({ completed: false }))
      const query$ = queryDb((get) => tables.todos.query.where(get(filter)).first({ fallback: () => defaultTodo }))

      expect(query$.run()).toMatchInlineSnapshot(`
      {
        "completed": false,
        "id": "",
        "text": "",
      }
    `)

      store.mutate(rawSqlMutation({ sql: sql`INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)` }))

      expect(query$.run()).toMatchInlineSnapshot(`
      {
        "completed": false,
        "id": "t1",
        "text": "buy milk",
      }
    `)

      query$.destroy()
      span.end()

      return { exporter }
    }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise)

    expect(getSimplifiedRootSpan(exporter)).toMatchSnapshot()
  })
})
