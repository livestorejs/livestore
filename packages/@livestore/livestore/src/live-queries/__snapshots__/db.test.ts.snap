// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`otel > otel 3`] = `
{
  "_name": "test",
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA journal_mode='MEMORY'; -- we don't flush to disk before committing a write
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "LiveStore:mutations",
      "children": [
        {
          "_name": "LiveStore:mutate",
          "attributes": {
            "livestore.mutateLabel": "mutate",
          },
          "children": [
            {
              "_name": "LiveStore:processWrites",
              "attributes": {
                "livestore.mutateLabel": "mutate",
              },
              "children": [
                {
                  "_name": "LiveStore:mutateWithoutRefresh",
                  "attributes": {
                    "livestore.args": "{
  "sql": "INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)"
}",
                    "livestore.mutation": "livestore.RawSql",
                  },
                  "children": [
                    {
                      "_name": "livestore.in-memory-db:execute",
                      "attributes": {
                        "sql.query": "INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)",
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "db:select * from todos",
          "attributes": {
            "sql.query": "select * from todos",
            "sql.rowsCount": 0,
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "select * from todos",
                "sql.rowsCount": 0,
              },
            },
          ],
        },
        {
          "_name": "db:select * from todos",
          "attributes": {
            "sql.query": "select * from todos",
            "sql.rowsCount": 1,
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "select * from todos",
                "sql.rowsCount": 1,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`otel > with thunks 3`] = `
{
  "_name": "test",
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA journal_mode='MEMORY'; -- we don't flush to disk before committing a write
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "LiveStore:mutations",
      "children": [
        {
          "_name": "LiveStore:mutate",
          "attributes": {
            "livestore.mutateLabel": "mutate",
          },
          "children": [
            {
              "_name": "LiveStore:processWrites",
              "attributes": {
                "livestore.mutateLabel": "mutate",
              },
              "children": [
                {
                  "_name": "LiveStore:mutateWithoutRefresh",
                  "attributes": {
                    "livestore.args": "{
  "sql": "INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)"
}",
                    "livestore.mutation": "livestore.RawSql",
                  },
                  "children": [
                    {
                      "_name": "livestore.in-memory-db:execute",
                      "attributes": {
                        "sql.query": "INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)",
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "db:select * from todos where completed = 0",
          "attributes": {
            "sql.query": "select * from todos where completed = 0",
            "sql.rowsCount": 0,
          },
          "children": [
            {
              "_name": "js:where-filter",
            },
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "select * from todos where completed = 0",
                "sql.rowsCount": 0,
              },
            },
          ],
        },
        {
          "_name": "db:select * from todos where completed = 0",
          "attributes": {
            "sql.query": "select * from todos where completed = 0",
            "sql.rowsCount": 1,
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "select * from todos where completed = 0",
                "sql.rowsCount": 1,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`otel > with thunks with query builder and without labels 3`] = `
{
  "_name": "test",
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA journal_mode='MEMORY'; -- we don't flush to disk before committing a write
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "LiveStore:mutations",
      "children": [
        {
          "_name": "LiveStore:mutate",
          "attributes": {
            "livestore.mutateLabel": "mutate",
          },
          "children": [
            {
              "_name": "LiveStore:processWrites",
              "attributes": {
                "livestore.mutateLabel": "mutate",
              },
              "children": [
                {
                  "_name": "LiveStore:mutateWithoutRefresh",
                  "attributes": {
                    "livestore.args": "{
  "sql": "INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)"
}",
                    "livestore.mutation": "livestore.RawSql",
                  },
                  "children": [
                    {
                      "_name": "livestore.in-memory-db:execute",
                      "attributes": {
                        "sql.query": "INSERT INTO todos (id, text, completed) VALUES ('t1', 'buy milk', 0)",
                      },
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "db:SELECT * FROM 'todos' WHERE completed = ? LIMIT ?",
          "attributes": {
            "sql.query": "SELECT * FROM 'todos' WHERE completed = ? LIMIT ?",
            "sql.rowsCount": 0,
          },
          "children": [
            {
              "_name": "js:() => ({ completed: false })",
            },
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT * FROM 'todos' WHERE completed = ? LIMIT ?",
                "sql.rowsCount": 0,
              },
            },
          ],
        },
        {
          "_name": "db:SELECT * FROM 'todos' WHERE completed = ? LIMIT ?",
          "attributes": {
            "sql.query": "SELECT * FROM 'todos' WHERE completed = ? LIMIT ?",
            "sql.rowsCount": 1,
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT * FROM 'todos' WHERE completed = ? LIMIT ?",
                "sql.rowsCount": 1,
              },
            },
          ],
        },
      ],
    },
  ],
}
`;
